### PR TITLE
#17460 - Fix and Improve Time per Task report

### PR DESCRIPTION
**Pull request by billy-editiano**
# Changes

## Summary
Add a command handler for generating TPT Report. This PR is basically moving the functionalities from datasaur-toolbox.

### Changes made from datasaur-toolbox
- GraphQL client setup is global in robosaur, reading the credentials from config file.
- Didn't use moment.js, using date-fns instead. As momentjs isn't maintained anymore.
- Using robosaur logger, replaced all console.log (the format is logstash by default)
- Improve output filename, includes date range and a base 32 timestamp 
  - to prevent the new report file replace other report files that already generated with same date range.
- Add `--outDir` options to specify where to save the report.

## Brief description of changed files
- Add generate-tpt-report command to commander program (`index.ts`)
- Generate TPT Report Handler, entrypoint for TPT report command (`src/handlers/generate-tpt-report.handler.ts`)
- Add 2 query helpers (`getProjectsWithAssignment` and `getRowAnalyticEvents`)
  - get projects with assignment (`src/datasaur/get-projects-with-assignment.ts`)
  - get row analytic events, to fetch all analytic event that this command will process. (`src/datasaur/get-row-analytic-events.ts`)
- Functions and Classes that related to the handler (`src/handlers/generate-tpt-report/`)
  - `ProjectCollection`, a helper class that manage and process project entities (`src/handlers/generate-tpt-report/project-collection.ts`)
  - `ReportBuilder`, a class that process events and calculate TPT (`src/handlers/generate-tpt-report/report-builder.ts`)

## How to test

1. Adjust credentials and URL in config file
  - create a json file `config.json`
  - adjust the content of the config to the following
    ```json
    {
      "datasaur": {
        "host": "https://app.datasaur.ai",
        "clientId": "your-client-id",
        "clientSecret": "your-client-secret"
      }
    }
    ```
  - you can get the `clientId` and `clientSecret` by refering to this [Guide](https://datasaurai.gitbook.io/datasaur/advanced/apis-docs/oauth-2.0#generate-oauth-credentials-menu)
  - the configuration will look like this
    ```json
    {
      "datasaur": {
        "host": "https://app.datasaur.ai",
        "clientId": "46752f68-aaaa-40f1-a988-e838fd94aaaa",
        "clientSecret": "5fa24715-aaaa-4b27-8c88-816ba5f25aaa"
      }
    }
    ```
2. run the following command
  a. with date range
    
    ```shell
    npm start -- generate-tpt-report <team-id> <path-to-config-file> \
    -s <start-date> \
    -e <end-date> \
    -o <path-to-a-directory>

    # example usage
    # create a folder for reports
    mkdir reports

    # run the command
    npm start -- generate-tpt-report 206 ./config.json \
    -s 2023-07-20 \
    -e 2023-07-27 \
    -o ./reports/
    ```
    
    b. all time
    
    ```shell
    npm start -- generate-tpt-report <team-id> <path-to-config-file> \
    --all \
    -o <path-to-a-directory>

    # example usage
    # create a folder for reports
    mkdir reports

    # run the command
    npm start -- generate-tpt-report 206 ./config.json \
    --all \
    -o ./reports/
    ```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "axios": "0.21.4",
     "cls-hooked": "^4.2.2",
     "commander": "7.2.0",
+    "date-fns": "^2.30.0",
```
Comment (tonoman3g) [inline package.json:23]: @billy-editiano boleh dijabarin changes yg berubah vs versi yang ada di datasaur-toolbox di PR ini apa aja?

taro di description PR aja

File: src/datasaur/get-projects-with-assignment.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  let data = await query(GET_PROJECT_WITH_ASSIGNMNET_QUERY, variables);
+  while (true) {
+    projects.push(...data.result.nodes);
+
+    if (data.result.pageInfo.nextCursor === null) {
+      break;
+    }
+
+    const nextVarialbes = {
```
Comment (tonoman3g) [inline src/datasaur/get-projects-with-assignment.ts:55]: typo, `nextVariables`

File: src/handlers/generate-tpt-report.handler.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  // generateAll or within a Date Range
+  if (all) {
+    await generateReportAllTime(teamId, projectCollection, outDir);
+  } else {
+    await generateReportWithinDateRange(teamId, projectCollection, startDate, endDate, outDir);
+  }
+}
+
+async function generateReportAllTime(teamId: string, projectCollection: ProjectCollection, outDir?: string) {
```
Comment (tonoman3g) [inline src/handlers/generate-tpt-report.handler.ts:58]: all timenya sebenernya misleading kah? karena ujung2nya max 7 hari?

File: src/index.ts [hunk 4]
```diff
@@ -54,6 +55,15 @@ program
   .description('Applies tags to projects based on the given config file')
   .action(handleApplyTags);
 
+program
+  .command(`generate-tpt-report <teamId> <configFile>`)
+  .option('-s, --startDate <value>', 'Start date in ISO 8601 format')
+  .option('-e, --endDate <value>', 'End date in ISO 8601 format')
+  .option('-o, --outDir <value>', 'Path to a directory where the output will be written')
+  .option('-a, --all', 'generate all-time TPT report', false)
+  .description('generate timestamp TPT report for a specific team with maximum 7 days period')
```
Comment (tonoman3g) [inline src/index.ts:64]: TPT nya dikasih kepanjangannya aja. Time Per Task

File: src/handlers/generate-tpt-report.handler.ts (hunk 3 above)
Comment (billy-editiano) [reply src/handlers/generate-tpt-report.handler.ts:58]: sbenernya all time kak, tapi dia get nya per 7 hari 🤔 berdasarkan code di `src/handlers/generate-tpt-report/helpers.ts`

Comment (tonoman3g) [reply src/handlers/generate-tpt-report.handler.ts:58]: oh baiklah

Comment (billy-editiano) [reply src/handlers/generate-tpt-report.handler.ts:58]: yang ini sih kak

```typescript
export const generateUnixTimestampRanges = (
  startTimestamp: number,
  endTimestamp: number,
  durationPerRange: number,
): UnixTimestampRange[] => {
  let currentTimestamp = startTimestamp;
  const result: UnixTimestampRange[] = [];

  while (currentTimestamp <= endTimestamp) {
    result.push({
      start: currentTimestamp,
      end: Math.min(currentTimestamp + durationPerRange, endTimestamp),
    });

    currentTimestamp += durationPerRange + 1;
  }

  return result;
};
```

jadi dia kaya nge bikin array of range per 7 hari wkwkwk

**Comment by tonoman3g** [comment]
waktu gw coba run, gw dpt error begini @billy-editiano 

```
{"@fields":{"command":"generate-tpt-report","error":"Cannot read properties of undefined (reading 'teamId')","error@context":{"_ns_name":"execution","id":-1,"trace-id":"c1ad990b-a4ca-4670-a8b7-5b323a61485c"},"level":"error","stack":"TypeError: Cannot read properties of undefined (reading 'teamId')\n    at setActiveTeamId (/Users/hartono/Workspace/robosaur/src/config/config.ts:65:27)\n    at setConfigByJSONFile (/Users/hartono/Workspace/robosaur/src/config/config.ts:38:3)\n    at _handleGenerateTptReport (/Users/hartono/Workspace/robosaur/src/handlers/generate-tpt-report.handler.ts:42:22)\n    at /Users/hartono/Workspace/robosaur/src/execution/handler-context.ts:40:44\n    at Namespace.runPromise (/Users/hartono/Workspace/robosaur/node_modules/cls-hooked/context.js:132:17)\n    at Command.<anonymous> (/Users/hartono/Workspace/robosaur/src/execution/handler-context.ts:19:31)\n    at Command.listener [as _actionHandler] (/Users/hartono/Workspace/robosaur/node_modules/commander/index.js:922:31)\n    at Command._parseCommand (/Users/hartono/Workspace/robosaur/node_modules/commander/index.js:1503:14)\n    at Command._dispatchSubcommand (/Users/hartono/Workspace/robosaur/node_modules/commander/index.js:1443:18)\n    at Command._parseCommand (/Users/hartono/Workspace/robosaur/node_modules/commander/index.js:1460:12)","teamId":"89","trace-id":"c1ad990b-a4ca-4670-a8b7-5b323a61485c"},"@message":"encountered unhandledRejection","@timestamp":"2023-07-27 16:43:57"}
```

**Review by tonoman3g — APPROVED** [review]

**Pull request merged**